### PR TITLE
Add more docs for CoerceUnsized and Unsize

### PR DESCRIPTION
**Pull request by Manishearth**
here be dragons

r? @ubsan @steveklabnik

### CODE REVIEW

File: src/libcore/marker.rs [hunk 1]
```diff
@@ -100,13 +100,26 @@ pub trait Sized {
 ///
 /// All implementations of `Unsize` are provided automatically by the compiler.
 ///
+/// `Unsize` is implemented for:
+///
+/// - `[T; N]` is `Unsize<[T]>`
+/// - `T` is `Trait` when `T: Trait`
```
Comment (ubsan) [inline src/libcore/marker.rs]: Should be `T` is `Unsize<Trait>` when `T: Trait`, in order to be consistent with the above sentence.

File: src/libcore/ops.rs [hunk 2]
```diff
@@ -2649,6 +2649,35 @@ mod impls {
 
 /// Trait that indicates that this is a pointer or a wrapper for one,
 /// where unsizing can be performed on the pointee.
+///
+/// See the [DST coercion RfC][dst-coerce] and [the nomicon entry on coercion][nomicon-coerce]
+/// for more details.
+///
+/// For builtin types, pointers to `T` will coerce to pointers to `U` if `T: Unsize<U>`
```
Comment (ubsan) [inline src/libcore/ops.rs]: "For builtin types" sounds unclear; perhaps reword this to say something to the effect of "for builtin pointer types".

File: src/libcore/ops.rs [hunk 3]
```diff
@@ -2649,6 +2649,35 @@ mod impls {
...
+/// provided an impl of `CoerceUnsized<Foo<U>> for Foo<T>` exists.
+/// Such an impl can only be written if `Foo<T>` has only a single non-phantomdata
+/// field involving `T`. If the type of that field is `Bar<T>`, an implementation
+/// of `CoerceUnsized<Bar<U>> for Bar<T>` must exist. The coercion will work by
+/// by coercing the `Bar<T>` field into `Bar<U>` and filling in the rest of the fields
+/// from `Foo<T>` to create a `Foo<U>`. This will effectively drill down to a pointer
+/// field and coerce that.
+///
+/// Generally, for smart pointers you will implement
+/// `CoerceUnsized<Ptr<U>> for Ptr<T> where T: Unsize<U>, U: ?Sized`, with an
```
Comment (ubsan) [inline src/libcore/ops.rs]: small nit; `where U: ?Sized` is, afaik, not valid syntax. Another argument for supporting it?

**Review by strega-nil — COMMENTED** [review]
Seems good on the technical aspects, otherwise.

File: src/libcore/ops.rs (hunk 3 above)
Comment (Manishearth) [inline src/libcore/ops.rs]: Yes. The docs also output this form. There is no reason not to support it IMO.

**Comment by Manishearth** [comment]
Updated.

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors: r+ rollup

**Comment by nrc** [comment]
cc @rust-lang/lang to make sure the docs fit our beliefs
